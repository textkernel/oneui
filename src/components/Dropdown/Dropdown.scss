@import '../../themes/oneui/oneui';

.Dropdown {
    &__list {
        background-color: var(--color-background);
        border-radius: var(--border-radius);
        border: var(--color-light) 1px solid;
        box-shadow: 0 5px 10px 0 rgba(0, 0, 0, 0.2);
        z-index: 400;

        .ListItem:hover:not(.ListItem--disabled) {
            background-color: var(--color-light-25);
        }
    }
}
