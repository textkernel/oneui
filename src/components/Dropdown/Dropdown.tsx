import React, { useState } from 'react';
import { useSelect } from 'downshift';
import { usePopper } from 'react-popper';
import { EmptyElement } from '../../customTypes/types';
import { bem } from '../../utils/bem/bem';
import { mergeRefs } from '../../utils/mergeRefs';
import { PopupPlacement } from '../../constants';
import { ButtonProps } from '../Buttons';
import { ListItemProps } from '../List/ListItem';
import { List } from '../List';
import styles from './Dropdown.scss';

const { elem } = bem('Dropdown', styles);

export interface Props<V> extends Omit<React.HTMLAttributes<HTMLDivElement>, 'onChange'> {
    /**
     *  Popup content to be placed inside List.
     *  Navigation available only if element has not empty `value` attribute
     *  and empty/false `disabled` attribute.
     *  Use {@link ListItem} component.
     */
    children:
        | React.ReactElement<ListItemProps>
        | (React.ReactElement<ListItemProps> | EmptyElement)[]
        | (
              | React.ReactElement<ListItemProps>
              | React.ReactElement<ListItemProps>[]
              | EmptyElement
          )[];
    /** {@link Button} element, controlled by current component */
    button: React.FunctionComponentElement<ButtonProps>;
    /**
     * Callback called on selecting one of the passed as children items.
     * Value parameter it is a `value` attribute of children item ({@link ListItemProps.value}).
     */
    onChange: (value: V) => void;
    /**
     * Callback called on toggle button click with current/previous open state
     */
    onToggleClick?: (prevOpenState: boolean) => void;
    /**
     * Callback called on focus of the dropdown menu
     */
    onMenuFocus?: () => void;
    /**
     * Callback called on blur of the dropdown menu
     */
    onMenuBlur?: () => void;
    /**
     * Additional select props that enrich and use downshift API
     */
    additionalSelectProps?: {};
    /**
     * ClassName that is assigned to <ul> element of the dropdown
     */
    listClassName?: string;
    /** Popup placement relative to button */
    placement?: PopupPlacement;
}

/**
 * ## Usage information
 * This component is intended for displaying available actions by button click.
 * It allows you to navigate with keyboard across non-disabled actions.
 * If you don't need navigation - use PopupBase component.

 * Navigation available only through children which have not empty `value` attribute
 * and empty/false `disabled` attribute (all other children items will be skipped during navigation).
 * Use `ListItem` component as child item.
 */

export function Dropdown<V>({
    button,
    children,
    onChange,
    onToggleClick = undefined,
    onMenuBlur = undefined,
    onMenuFocus = undefined,
    placement = 'bottom-end',
    additionalSelectProps = {},
    listClassName = '',
    ...rest
}: Props<V>) {
    const [referenceElement, setReferenceElement] = useState(null);
    const [popperElement, setPopperElement] = useState(null);

    const state = usePopper(referenceElement, popperElement, {
        placement,
        modifiers: [
            {
                name: 'offset',
                options: { offset: [0, 3] },
            },
        ],
    });

    const childrenArray = React.Children.toArray(children);
    const valuesAvailableForHighlight: V[] = [];
    childrenArray.forEach((child) => {
        if (
            React.isValidElement(child) &&
            child.props.value !== undefined &&
            child.props.value !== null &&
            !child.props.disabled
        ) {
            valuesAvailableForHighlight.push(child.props.value);
        }
    });

    const { isOpen, getItemProps, getMenuProps, getToggleButtonProps, highlightedIndex, reset } =
        useSelect<V>({
            items: valuesAvailableForHighlight,
            onSelectedItemChange: ({ selectedItem }) => {
                if (selectedItem) {
                    onChange(selectedItem);
                    reset();
                }
            },
            ...additionalSelectProps,
        });

    const menuProps = getMenuProps();
    const toggleButtonProps = getToggleButtonProps({
        onClick: () => {
            onToggleClick?.(isOpen);
        },
        onBlur: onMenuBlur,
        onFocus: onMenuFocus,
    });
    const openPopperProps = isOpen && {
        style: state.styles.popper,
        ...state.attributes.popper,
        ...elem('list', {
            elemClassName: listClassName,
        }),
    };
    return (
        <>
            {React.cloneElement<any>(button, {
                ...rest,
                ...toggleButtonProps,
                ref: mergeRefs([setReferenceElement, toggleButtonProps.ref, button.ref]),
            })}
            <List
                {...menuProps}
                {...openPopperProps}
                ref={mergeRefs([isOpen ? setPopperElement : undefined, menuProps.ref])}
                isControlledNavigation
            >
                {isOpen &&
                    childrenArray.map((child) => {
                        if (
                            React.isValidElement(child) &&
                            valuesAvailableForHighlight.includes(child.props.value)
                        ) {
                            const currentValueIndex = valuesAvailableForHighlight.findIndex(
                                (val) => val === child.props.value
                            );

                            return React.cloneElement<any>(child, {
                                ...getItemProps({
                                    index: currentValueIndex,
                                    item: child.props.value,
                                }),
                                isHighlighted: highlightedIndex === currentValueIndex,
                            });
                        }
                        return child;
                    })}
            </List>
        </>
    );
}

Dropdown.displayName = 'Dropdown';
