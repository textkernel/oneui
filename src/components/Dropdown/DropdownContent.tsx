import React, { useEffect } from 'react';
import { Content, DropdownMenuContentProps } from '@radix-ui/react-dropdown-menu';
import { bem } from '../../utils/bem/bem';
import styles from './Dropdown.scss';

const { block } = bem('DropdownContent', styles);

interface Props extends DropdownMenuContentProps {
    refElement?: React.RefObject<HTMLElement | null>;
}

export const DropdownContent = ({ refElement, ...rest }: Props) => {
    const [width, setWidth] = React.useState<number>();

    useEffect(() => {
        if (refElement?.current?.offsetWidth) {
            setWidth(refElement.current.offsetWidth);
        }
    }, [refElement]);

    return <Content {...rest} {...block(rest)} align="start" style={{ width }} />;
};
