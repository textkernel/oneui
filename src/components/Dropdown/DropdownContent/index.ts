export { DropdownContent } from './DropdownContent';
export { DropdownMenuContentProps as DropdownContentProps } from '@radix-ui/react-dropdown-menu';
