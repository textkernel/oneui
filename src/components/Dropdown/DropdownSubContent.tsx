import React from 'react';
import { SubContent, DropdownMenuSubContentProps } from '@radix-ui/react-dropdown-menu';
import { bem } from '../../utils/bem/bem';
import styles from './Dropdown.scss';

const { block } = bem('DropdownSubContent', styles);

export const DropdownSubContent = (props: DropdownMenuSubContentProps) => (
    <SubContent {...props} {...block(props)} sideOffset={6} />
);
