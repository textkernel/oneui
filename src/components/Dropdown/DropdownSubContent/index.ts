export { DropdownSubContent } from './DropdownSubContent';
export { DropdownMenuSubContentProps as DropdownSubContentProps } from '@radix-ui/react-dropdown-menu';
