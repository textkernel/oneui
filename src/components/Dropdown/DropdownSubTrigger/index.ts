export { DropdownSubTrigger } from './DropdownSubTrigger';
export { DropdownMenuSubTriggerProps as DropdownSubTriggerProps } from '@radix-ui/react-dropdown-menu';
