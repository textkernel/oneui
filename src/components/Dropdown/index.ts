export { Dropdown, Props as DropdownProps } from './Dropdown';
