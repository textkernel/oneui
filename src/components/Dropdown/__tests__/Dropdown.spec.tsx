import React from 'react';
import { screen, render, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Dropdown } from '../Dropdown';
import { Button } from '../../Buttons';
import { ListItem } from '../../List';
import '@testing-library/jest-dom';

describe('Dropdown', () => {
    const mockOnChange = jest.fn();
    const mockOnButtonClick = jest.fn();
    const mockOnMenuFocus = jest.fn();
    const mockOnMenuBlur = jest.fn();
    const mockOnDropdownStateChange = jest.fn();
    let view;

    const getInteractiveListItemAt = (index) => screen.getAllByRole('option')[index];

    beforeEach(() => {
        view = render(
            <Dropdown
                button={<Button isPrimary>Click me!</Button>}
                onChange={mockOnChange}
                placement="top-start"
                onToggleClick={mockOnButtonClick}
                onMenuFocus={mockOnMenuFocus}
                onMenuBlur={mockOnMenuBlur}
                additionalSelectProps={{
                    onStateChange: mockOnDropdownStateChange,
                }}
            >
                <ListItem key="disabled-key" disabled>
                    Disabled
                </ListItem>
                <ListItem key="first-key" value="first-value">
                    With value
                </ListItem>
            </Dropdown>
        );
    });

    it('should render correctly closed', () => {
        expect(view.container).toMatchSnapshot();
        expect(screen.queryAllByRole('option')).toHaveLength(0);
        expect(screen.queryAllByRole('listitem')).toHaveLength(0);
    });

    it('should render correctly opened', async () => {
        const user = userEvent.setup();
        await user.click(screen.getByRole('button', { name: 'Click me!' }));

        expect(view.container).toMatchSnapshot();
        expect(screen.getAllByRole('listbox')).toHaveLength(1);
        expect(screen.getAllByRole('option')).toHaveLength(1);
        expect(screen.getAllByRole('listitem')).toHaveLength(1);
    });
    it('should access only enabled items with value during keyboard navigation', async () => {
        const user = userEvent.setup();
        view.rerender(
            <Dropdown
                button={<Button isPrimary>Click me!</Button>}
                onChange={mockOnChange}
                placement="top-start"
            >
                <ListItem key="key-1" disabled>
                    Disabled
                </ListItem>
                <ListItem key="key-2" value="1">
                    With value 1
                </ListItem>
                <ListItem key="key-3">Without value</ListItem>
                <div className="customDiv" key="key-4">
                    Div
                </div>
                <ListItem key="key-5" value="2">
                    With value 2
                </ListItem>
            </Dropdown>
        );
        await user.click(screen.getByRole('button', { name: 'Click me!' }));
        expect(view.container).toMatchSnapshot();
        // focus list
        await user.click(screen.getByRole('listbox'));
        const navigateDown = async () => user.keyboard('[ArrowDown]');

        // 1 keydown
        await navigateDown();
        expect(getInteractiveListItemAt(0).getAttribute('aria-selected')).toBeTruthy();
        expect(getInteractiveListItemAt(0)).toHaveTextContent('With value 1');
        // 2 keydown
        await navigateDown();
        expect(getInteractiveListItemAt(1).getAttribute('aria-selected')).toBeTruthy();
        expect(getInteractiveListItemAt(1)).toHaveTextContent('With value 2');
        // 3 keydown. Should be again last not disabled with value item => 'With value 2'
        await navigateDown();
        expect(getInteractiveListItemAt(1).getAttribute('aria-selected')).toBeTruthy();
        expect(getInteractiveListItemAt(1)).toHaveTextContent('With value 2');
    });

    it('onChange should return passed value', async () => {
        const user = userEvent.setup();
        view.rerender(
            <Dropdown
                button={<Button isPrimary>Click me!</Button>}
                onChange={mockOnChange}
                placement="top-start"
            >
                <ListItem key="key-1" disabled>
                    Disabled
                </ListItem>
                <ListItem key="key-2" value="testValue">
                    With value
                </ListItem>
            </Dropdown>
        );
        await user.click(screen.getByRole('button', { name: 'Click me!' }));

        await user.keyboard('[ArrowDown]');
        await user.keyboard('[Enter]');

        expect(mockOnChange).toBeCalledWith('testValue');
    });

    it('should render correctly with mixed children: array and single ListItem', async () => {
        const user = userEvent.setup();
        view.rerender(
            <Dropdown
                button={<Button isPrimary>Click me!</Button>}
                onChange={mockOnChange}
                placement="top-start"
            >
                <ListItem key="disabled-key" disabled>
                    Disabled
                </ListItem>
                {['one', 'two'].map((value) => (
                    <ListItem key={value} value={value}>
                        {value}
                    </ListItem>
                ))}
            </Dropdown>
        );
        await user.click(screen.getByRole('button', { name: 'Click me!' }));
        expect(view.container).toMatchSnapshot();
        expect(screen.getAllByRole('listitem')).toHaveLength(1);
        expect(screen.getAllByRole('option')).toHaveLength(2);
    });

    it('should call cb when button is clicked', async () => {
        const user = userEvent.setup();
        await user.click(screen.getByRole('button', { name: 'Click me!' }));
        expect(mockOnButtonClick).toHaveBeenCalledWith(false);
        await user.click(screen.getByRole('button', { name: 'Click me!' }));
        expect(mockOnButtonClick).toHaveBeenCalledWith(true);
    });

    it('should call callback when menu is focused', async () => {
        const user = userEvent.setup();
        await user.click(screen.getByRole('button', { name: 'Click me!' }));
        await user.click(screen.getByRole('listbox'));

        expect(mockOnMenuFocus).toHaveBeenCalled();
    });

    it('should call callback when menu is blurred', async () => {
        const user = userEvent.setup();
        await user.click(screen.getByRole('button', { name: 'Click me!' }));
        await user.click(screen.getByRole('listbox'));
        await user.click(document.body);

        expect(mockOnMenuBlur).toHaveBeenCalled();
    });

    it('should call callback when menu state is changed', async () => {
        const user = userEvent.setup();
        await user.click(screen.getByRole('button', { name: 'Click me!' }));
        // Event for a successful mouse click on dropdown button
        expect(mockOnDropdownStateChange).toHaveBeenCalledWith({
            isOpen: true,
            type: '__togglebutton_click__',
        });

        await user.keyboard('[ArrowDown]');
        // Event for a successful arrow down press when menu is opened
        expect(mockOnDropdownStateChange).toHaveBeenCalledWith({
            highlightedIndex: 0,
            type: '__menu_keydown_arrow_down__',
        });

        await user.click(document.body);
        // Event for a successful dropdown blur
        expect(mockOnDropdownStateChange).toHaveBeenCalledWith({
            highlightedIndex: -1,
            isOpen: false,
            type: '__menu_blur__',
        });
    });

    it('should open dropdown by default if corresponding prop is set', async () => {
        view = render(
            <Dropdown
                button={<Button isPrimary>Click me!</Button>}
                onChange={mockOnChange}
                placement="top-start"
                additionalSelectProps={{
                    initialIsOpen: true,
                }}
            >
                <ListItem key="disabled-key" disabled>
                    Disabled
                </ListItem>
                {['one', 'two'].map((value) => (
                    <ListItem key={value} value={value}>
                        {value}
                    </ListItem>
                ))}
            </Dropdown>
        );
        // we need to wait for the popup state to finish updating
        await waitFor(() => screen.getByRole('listitem'));

        expect(screen.getAllByRole('listitem')).toHaveLength(1);
        expect(screen.getAllByRole('option')).toHaveLength(2);
    });

    it('should allow for conditional rendering of items', async () => {
        const user = userEvent.setup();
        const condition = false;
        view.rerender(
            <Dropdown
                button={<Button isPrimary>Click me!</Button>}
                onChange={mockOnChange}
                placement="top-start"
            >
                {condition ? (
                    <ListItem key="key-1" disabled>
                        Disabled
                    </ListItem>
                ) : null}
                <ListItem key="key-2" value="testValue">
                    With value
                </ListItem>
                {condition ? <ListItem key="3">should not render</ListItem> : null}
            </Dropdown>
        );
        await user.click(screen.getAllByRole('button', { name: 'Click me!' })[1]);

        expect(view.container).toMatchSnapshot();
        expect(screen.queryByText('false')).not.toBeInTheDocument();
    });
});
