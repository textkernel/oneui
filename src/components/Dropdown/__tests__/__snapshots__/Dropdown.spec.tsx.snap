// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown should access only enabled items with value during keyboard navigation 1`] = `
<div>
  <button
    aria-expanded="true"
    aria-haspopup="listbox"
    aria-labelledby="downshift-11-label downshift-11-toggle-button"
    class="Button Button--isPrimary"
    id="downshift-11-toggle-button"
    type="button"
  >
    Click me!
  </button>
  <ul
    aria-labelledby="downshift-11-label"
    class="List Dropdown__list"
    data-popper-escaped="true"
    data-popper-placement="top-start"
    data-popper-reference-hidden="true"
    id="downshift-11-menu"
    role="listbox"
    style="position: absolute; left: 0px; top: 0px; bottom: 0px; transform: translate(0px, -3px);"
    tabindex="-1"
  >
    <li
      class="ListItem ListItem--disabled List__item"
    >
      <span>
        Disabled
      </span>
    </li>
    <li
      aria-selected="false"
      class="ListItem ListItem--clickable List__item"
      id="downshift-11-item-0"
      role="option"
    >
      <span>
        With value 1
      </span>
    </li>
    <li
      class="ListItem List__item"
    >
      <span>
        Without value
      </span>
    </li>
    <div
      class="List__item customDiv"
    >
      Div
    </div>
    <li
      aria-selected="false"
      class="ListItem ListItem--clickable List__item"
      id="downshift-11-item-1"
      role="option"
    >
      <span>
        With value 2
      </span>
    </li>
  </ul>
</div>
`;

exports[`Dropdown should allow for conditional rendering of items 1`] = `
<div>
  <button
    aria-expanded="false"
    aria-haspopup="listbox"
    aria-labelledby="downshift-107-label downshift-107-toggle-button"
    class="Button Button--isPrimary"
    id="downshift-107-toggle-button"
    type="button"
  >
    Click me!
  </button>
  <ul
    aria-labelledby="downshift-107-label"
    class="List"
    id="downshift-107-menu"
    role="listbox"
    tabindex="-1"
  />
</div>
`;

exports[`Dropdown should render correctly closed 1`] = `
<div>
  <button
    aria-expanded="false"
    aria-haspopup="listbox"
    aria-labelledby="downshift-0-label downshift-0-toggle-button"
    class="Button Button--isPrimary"
    id="downshift-0-toggle-button"
    type="button"
  >
    Click me!
  </button>
  <ul
    aria-labelledby="downshift-0-label"
    class="List"
    id="downshift-0-menu"
    role="listbox"
    tabindex="-1"
  />
</div>
`;

exports[`Dropdown should render correctly opened 1`] = `
<div>
  <button
    aria-expanded="true"
    aria-haspopup="listbox"
    aria-labelledby="downshift-3-label downshift-3-toggle-button"
    class="Button Button--isPrimary"
    id="downshift-3-toggle-button"
    type="button"
  >
    Click me!
  </button>
  <ul
    aria-labelledby="downshift-3-label"
    class="List Dropdown__list"
    data-popper-escaped="true"
    data-popper-placement="top-start"
    data-popper-reference-hidden="true"
    id="downshift-3-menu"
    role="listbox"
    style="position: absolute; left: 0px; top: 0px; bottom: 0px; transform: translate(0px, -3px);"
    tabindex="-1"
  >
    <li
      class="ListItem ListItem--disabled List__item"
    >
      <span>
        Disabled
      </span>
    </li>
    <li
      aria-selected="false"
      class="ListItem ListItem--clickable List__item"
      id="downshift-3-item-0"
      role="option"
    >
      <span>
        With value
      </span>
    </li>
  </ul>
</div>
`;

exports[`Dropdown should render correctly with mixed children: array and single ListItem 1`] = `
<div>
  <button
    aria-expanded="true"
    aria-haspopup="listbox"
    aria-labelledby="downshift-44-label downshift-44-toggle-button"
    class="Button Button--isPrimary"
    id="downshift-44-toggle-button"
    type="button"
  >
    Click me!
  </button>
  <ul
    aria-labelledby="downshift-44-label"
    class="List Dropdown__list"
    data-popper-escaped="true"
    data-popper-placement="top-start"
    data-popper-reference-hidden="true"
    id="downshift-44-menu"
    role="listbox"
    style="position: absolute; left: 0px; top: 0px; bottom: 0px; transform: translate(0px, -3px);"
    tabindex="-1"
  >
    <li
      class="ListItem ListItem--disabled List__item"
    >
      <span>
        Disabled
      </span>
    </li>
    <li
      aria-selected="false"
      class="ListItem ListItem--clickable List__item"
      id="downshift-44-item-0"
      role="option"
    >
      <span>
        one
      </span>
    </li>
    <li
      aria-selected="false"
      class="ListItem ListItem--clickable List__item"
      id="downshift-44-item-1"
      role="option"
    >
      <span>
        two
      </span>
    </li>
  </ul>
</div>
`;
